Elements/test/AdaptiveGraphRoutingTests.cs

```diff
@@ -432,6 +578,100 @@ public void AdaptiveGraphRoutingSimpleTreeExample()
+                new Vector3(9, 8, 0),
+                new Vector3(11, 12, 0),
+                new Vector3(12, 20, 0),
+            };
+
+            //4. All significant points must be added as key points.
+            var keyPoints = new List<Vector3>();
+            keyPoints.AddRange(inputPoints);
+
+            //5. Define hint and offset lines.
+            var firstOffsetPolyline = new Polyline(new List<Vector3>(){
+                new Vector3(5, 2),
+                new Vector3(5, 20)
+            });
+            keyPoints.AddRange(firstOffsetPolyline.Vertices);
+
+            var secondOffsetPolyline = new Polyline(new List<Vector3>()
+            {
+                new Vector3(15, 2),
+                new Vector3(15, 20)
+            });
+            keyPoints.AddRange(secondOffsetPolyline.Vertices);
+
+            var hintPolyline = new Polyline(new List<Vector3>()
+            {
+                new Vector3(5, 17),
+                new Vector3(15, 17),
+
+            });
+            keyPoints.AddRange(hintPolyline.Vertices);
+
+            //6. Define obstacles.
+            var box = new BBox3(new Vector3(11, 14, 0), new Vector3(17, 18, 0));
+            var obstacle = Obstacle.FromBBox(box);
+            keyPoints.AddRange(box.Corners());
+
+            //7. Create grid.
+            AdaptiveGrid grid = new AdaptiveGrid();
+            grid.AddFromPolygon(boundary, keyPoints);
+            grid.SubtractObstacle(obstacle);
+
+            //8. Get indices's for start and end vertices
+            var inputVertices = new List<RoutingVertex>();
+            foreach (var input in inputPoints)
+            {
+                Assert.True(grid.TryGetVertexIndex(input, out ulong id, grid.Tolerance));
+                {
+                    inputVertices.Add(new RoutingVertex(id, 0.5));
+                }
+            }
+
+            //9. Set configurations for hint and offset lines.
+            var hint = new RoutingHintLine(hintPolyline,
+                factor: 0.01, influence: 0.1, userDefined: true, is2D: true);
+            var offset1 = new RoutingHintLine(firstOffsetPolyline,
+                factor: 0.9, influence: 0.1, userDefined: false, is2D: true);
+            var offset2 = new RoutingHintLine(secondOffsetPolyline,
+                factor: 0.9, influence: 0.1, userDefined: false, is2D: true);
+            var hints = new List<RoutingHintLine> { hint, offset1, offset2 };
+
+            //10. Run algorithm
+            var config = new RoutingConfiguration(turnCost: 1);
+            AdaptiveGraphRouting alg = new AdaptiveGraphRouting(grid, config);
+            ulong tailVertex = 0;
+            var tree = alg.BuildSteinerTree(inputVertices, hints);
+
+            //Not throws if no hint lines
+            alg.BuildSteinerTree(inputVertices, new List<RoutingHintLine>());
+
+            //Results visualization
+            VisualizeRoutingTree(grid, inputVertices, tree);
+            VisualizeGrid(alg, hints, keyPoints);
```

Comment:
I've tried a test 
[Theory]
        [InlineData(true)]
        [InlineData(false)]
        public void AdaptiveGridRoutingBestIndividualRoute(bool old)
        {
            var inputs = new List<Vector3>
            {
                (-5, 10), (4, 5), (2, 12)
            };

            var lines = new List<List<Vector3>>()
            {
                new List<Vector3>{ (0, 0), (0, 10) },
                new List<Vector3>{ (-5, 10), (2, 10) },
                new List<Vector3>{ (2, 12), (2, 5) },
                new List<Vector3>{ (4, 5), (0, 5) }
            };

            var grid = new AdaptiveGrid();
            foreach (var line in lines)
            {
                grid.AddVertices(line, AdaptiveGrid.VerticesInsertionMethod.ConnectAndCut);
            }

            var inputVertices = inputs.Select(i =>
            {
                Assert.True(grid.TryGetVertexIndex(i, out var id, grid.Tolerance));
                return new RoutingVertex(id, 0);
            }).ToList();

            Assert.True(grid.TryGetVertexIndex((0, 0), out var end, grid.Tolerance));

            var alg = new AdaptiveGraphRouting(grid, new RoutingConfiguration(turnCost: 1));
            var hints = new List<RoutingHintLine>();

            IDictionary<ulong, TreeNode> tree;
            if (old)
                tree = alg.BuildSpanningTree(inputVertices, end, hints, TreeOrder.ClosestToFurthest);
            else
                tree = alg.BuildSteinerTree(inputVertices.Append(new RoutingVertex(end, 0)).ToList(), hints);

            CheckTree(grid, inputVertices[0].Id, tree,
                      new List<Vector3> { (-5, 10), (0, 10), (0, 5), (0, 0) });
            CheckTree(grid, inputVertices[1].Id, tree,
                      new List<Vector3> { (4, 5), (2, 5), (0, 5), (0, 0) });
            CheckTree(grid, inputVertices[2].Id, tree,
                      new List<Vector3> { (2, 12), (2, 10), (0, 10), (0, 5), (0, 0) });
        }
It shows value of being connected to the infrastructure. 
But it doesn't work , looks like your links are connected backwards and trunk is null for the first point.
I think we should still have defined trunk parameter in the function and make sure that all links are directed correctly. Let's focus on that. Also I would like you to try do similar things with other tests that use single list of leafs. Let's focus on this for now.